File: 0820-find-eventual-safe-states/0820-find-eventual-safe-states.java
class Solution {
    public List<Integer> eventualSafeNodes(int[][] graph) {
        Queue<Integer> q = new LinkedList<>();
        int n= graph.length;
        ArrayList<ArrayList<Integer>> reverseGraph = new ArrayList<>();
        ArrayList<Integer> topo = new ArrayList<>();
        int outd[] = new int[n];
        
        for(int i=0;i<n;i++)reverseGraph.add(new ArrayList<>());

        for(int i=0;i<n;i++){
            for(int it :graph[i])
                reverseGraph.get(it).add(i);
            outd[i] = graph[i].length;
            if(outd[i] == 0)
            q.add(i);
        }

        while(!q.isEmpty()){
            int rem = q.poll();
            topo.add(rem);

            for(int it:reverseGraph.get(rem)){
                outd[it]--;
                if(outd[it] == 0)
                q.add(it);
            }
        }
        Collections.sort(topo);
        return topo;
    }
}

File: 0820-find-eventual-safe-states/0820-find-eventual-safe-states.cpp
class Solution {
public:
    vector<int> eventualSafeNodes(vector<vector<int>>& graph) {
        int V = graph.size();
        vector<vector<int>> adj(V);
        vector<int> indeg(V, 0);
        for(int i=0;i<V;i++){
            for(auto it:graph[i]){
                adj[it].push_back(i);
                indeg[i]++;
            }
        }

        queue<int> q;
        vector<int> ans;

        for(int i=0;i<V;i++)
            if(indeg[i] == 0)q.push(i);

        while(!q.empty()){
            int node = q.front();
            q.pop();
            ans.push_back(node);

            for(auto it:adj[node]){
                indeg[it]--;
                if(indeg[it] == 0)q.push(it);
            }
        }
        sort(ans.begin(), ans.end());
        return ans;
    }
};